File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_11/java/Solution.java
package com.hi.dhl.algorithms.offer._11.java;

/**
 * <pre>
 *     author: dhl
 *     desc  :
 * </pre>
 */
class Solution {
    public int minArray(int[] numbers) {

        int left = 0;
        int right = numbers.length - 1;

        while (left < right) {
            int mid = (left + right) >>> 2;
            if (numbers[mid] < numbers[right]) {
                right = mid;
            } else if (numbers[mid] > numbers[right]) {
                left = mid + 1;
            } else {
                right = right - 1;
            }
        }

        return numbers[left];
    }

    public static void main(String... args) {
        Solution solution = new Solution();
        System.out.println(solution.minArray(new int[]{2, 2, 2, 0, 1}));
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_11/kotlin/Solution.kt
package com.hi.dhl.algorithms.offer._11.kotlin

/**
 * <pre>
 *     author: dhl
 *     desc  :
 * </pre>
 */

class Solution {
    fun minArray(numbers: IntArray): Int {
        var left = 0
        var right = numbers.size - 1
        while (left < right) {
            val mid = (left + right) ushr 1
            when {
                numbers[mid] > numbers[right] -> left = mid + 1
                numbers[mid] < numbers[right] -> right = mid
                else -> right = right - 1
            }
        }
        return numbers[left]
    }
}

fun main() {
    val solution = Solution()
    println(solution.minArray(intArrayOf(1, 1)))
}